### PR TITLE
[mh-248] enable long emails, email-based authentication

### PR DESCRIPTION
**Pull request by sharrisoncaktus**
We have had a long-standing bug that arises from the fact that user.username was being set to user.email during User save (via pre-save receiver).  But User.username can only be 30 characters long. So, any email longer than 30 characters was causing 500 errors during signup and edit profile.

The solution is twofold: 

1. When saving the User, the username is now a hash of the email that is algorithmically limited to 28 characters (base64 encoded SHA-1 hash digest) while still giving a reasonable guarantee of uniqueness (although SHA-1 hash collisions do occur on very rare occasion — something I haven't accounted for).

2. Far larger a change, I've switched the system to email-based authentication by creating a custom EmailAuthBackend (in myhpom.auth) and assigning it in settings. This also affects multiple places in the app: Every test that uses Client.login() with username had to be switched to email. So most of the changes are of that sort.

In the end, this gets us part of the way toward our goal of creating a custom user model that uses email-based authentication.

### CODE REVIEW

File: myhpom/auth.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from django.contrib.auth import get_user_model
+from django.contrib.auth.backends import ModelBackend
+
+class EmailAuthBackend(ModelBackend):
```
Comment (dsummersl) [inline myhpom/auth.py]: pep8 would say two spaces between imports and classes (I'm sorry for this - we should really get a linter into our CI pipeline).

File: myhpom/auth.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    Authenticate a user via email (which is unique in this system) rather than username.
+    """
+    def authenticate(self, username=None, password=None, **kwargs):
+        User = get_user_model()
+        try:
+            user = User.objects.get(email=username)
+        except User.DoesNotExist:
+            return None
+        else:
+            if user.check_password(password):
```
Comment (dsummersl) [inline myhpom/auth.py]: Add `and self.user_can_authenticate(user)` to prevent inactive users from logging in.

File: myhpom/models/user.py [hunk 3]
```diff
@@ -1,8 +1,22 @@
 from django.db import models
 from django.contrib.auth.models import User
 from myhpom import validators
+import base64, hashlib
 
-"""Connect a pre_save receiver/hook to the User model, and use this model in all imports"""
+"""
+* Connect a pre_save receiver/hook to the User model, and use this model in all imports
```
Comment (dsummersl) [inline myhpom/models/user.py]: gah another e302 - I installed flake8 in a hydroshare instance and ran just to be sure i'm not enforcing some rando rule on my system, and its there too:

```
root@3b657ed24655:/hydroshare# flake8 myhpom/models/user.py
myhpom/models/user.py:4:14: E401 multiple imports on one line
myhpom/models/user.py:13:1: E302 expected 2 blank lines, found 1
myhpom/models/user.py:14:41: W291 trailing whitespace
```

File: myhpom/models/user.py [hunk 4]
```diff
@@ -1,8 +1,22 @@
...
+* monkeypatch User.__unicode__ to return the email address
+"""
+
+User.__unicode__ = lambda u: u.email
+
+def get_username(email):
+    """unique username as hash of the email. 
+    Must be <= 30 characters long. 30, Django? Really?
```
Comment (dsummersl) [inline myhpom/models/user.py]: django 2.1 source shows they bumped it up to 150. But in antiquated 1.8: really.

**Review by dsummersl — CHANGES_REQUESTED** [review]
Only real issue that needs to be improved is to include a `user_can_authenticate()` check in the authentication model.

File: myhpom/models/user.py (hunk 3 above)
Comment (sharrisoncaktus) [reply myhpom/models/user.py]: well at least my updated flake8 agrees with yours!

Comment (dsummersl) [reply myhpom/models/user.py]: woot!

**Review by dsummersl — APPROVED** [review]
✉️

**Pull request merged**